onnxruntime/core/framework/cloud_invoker.cc

```diff
@@ -33,6 +36,174 @@ CloudEndPointInvoker::CloudEndPointInvoker(const CloudEndPointConfig& config,
+  }
+  ORT_DISALLOW_COPY_ASSIGNMENT_AND_MOVE(CurlGlobal);
+};
+
+// OpenAIInvoker
+class OpenAIInvoker : public CloudEndPointInvoker {
+ public:
+  OpenAIInvoker(const CloudEndPointConfig& config, const AllocatorPtr& allocator);
+  onnxruntime::Status Send(const CloudEndPointConfig& run_options,
+                           const InlinedVector<std::string>& input_names,
+                           gsl::span<const OrtValue> ort_inputs,
+                           const InlinedVector<std::string>& output_names,
+                           std::vector<OrtValue>& ort_outputs) const override;
+
+ private:
+  std::string uri_;
+  std::string model_name_;
+};
+
+OpenAIInvoker::OpenAIInvoker(const CloudEndPointConfig& config,
+                             const AllocatorPtr& allocator) : CloudEndPointInvoker(config, allocator) {
+  CurlGlobal::Init();
+  ReadConfig(kAzureUri, uri_);
+  ReadConfig(kAzureModelName, model_name_);
+}
+
+struct StringBuffer {
+  StringBuffer() = default;
+  ~StringBuffer() = default;
+  ORT_DISALLOW_COPY_ASSIGNMENT_AND_MOVE(StringBuffer);
+  std::stringstream ss_;
+};
+
+// apply the callback only when response is for sure to be a '/0' terminated string
+static size_t WriteStringCallback(void* contents, size_t size, size_t nmemb, void* userp) {
+  size_t realsize = size * nmemb;
+  auto buffer = reinterpret_cast<struct StringBuffer*>(userp);
+  buffer->ss_.write(reinterpret_cast<const char*>(contents), realsize);
+  return realsize;
+}
+
+using CurlWriteCallBack = size_t (*)(void*, size_t, size_t, void*);
+
+class CurlHandler {
+ public:
+  CurlHandler(CurlWriteCallBack call_back) {
+    curl_ = curl_easy_init();
+    curl_easy_setopt(curl_, CURLOPT_BUFFERSIZE, 102400L);
+    curl_easy_setopt(curl_, CURLOPT_NOPROGRESS, 1L);
+    curl_easy_setopt(curl_, CURLOPT_USERAGENT, "curl/7.83.1");
+    curl_easy_setopt(curl_, CURLOPT_MAXREDIRS, 50L);
+    curl_easy_setopt(curl_, CURLOPT_FTP_SKIP_PASV_IP, 1L);
+    curl_easy_setopt(curl_, CURLOPT_TCP_KEEPALIVE, 1L);
+    curl_easy_setopt(curl_, CURLOPT_WRITEFUNCTION, call_back);
+  }
+  ~CurlHandler() {
+    if (curl_) {
+      curl_easy_cleanup(curl_);
+      curl_ = {};
+    }
+    if (mime1_) {
+      curl_mime_free(mime1_);
+      mime1_ = {};
+    }
+    if (headers_) {
+      curl_slist_free_all(headers_);
+      headers_ = {};
+    }
+    if (from_) {
+      curl_formfree(from_);
+      from_ = {};
+    }
+  }
+  void AddHeader(const char* data) {
+    headers_ = curl_slist_append(headers_, data);
+  }
+  template <typename... Args>
+  void AddForm(Args... args) {
+    curl_formadd(&from_, &last_, args...);
+  }
+  template <typename T>
+  void SetOption(CURLoption opt, T val) {
+    curl_easy_setopt(curl_, opt, val);
+  }
+  ORT_DISALLOW_COPY_ASSIGNMENT_AND_MOVE(CurlHandler);
+  CURLcode Perform() {
+    SetOption(CURLOPT_HTTPHEADER, headers_);
+    SetOption(CURLOPT_HTTPPOST, from_);
+    return curl_easy_perform(curl_);
+  }
+
+ private:
+  CURL* curl_{};
```

Comment:
You may change it to a std::unique_ptr.  See https://stackoverflow.com/questions/27440953/stdunique-ptr-for-c-functions-that-need-free for an example.  Instead of using std::free to free the memory, in your case you need to use `curl_easy_cleanup`

Reply:
Agree ... we got to call curl_easy_cleanup whatsoever in destructor.

Reply:
We see  std::unique_ptr is safer than raw pointers, which means your current code is not wrong, but using std::unique_ptr could make the code easier to read and verify.  Just a suggestion. You don't have to take it.

Reply:
Done.